### ISSUE TITLE
loss definition

### ISSUE THREAD

**Issue by legolas140** [michalfaber/keras_Realtime_Multi-Person_Pose_Estimation#9]
Hello. I am new to this program. When I read the code in train_pose.py, I am confused about the definition of the losses. It doesn't seem to reflect the definition of the Part Detection losses and Part Affinity Fields losses in the paper. Could you tell me where to find this part? Thank you very much!!

**Comment by michalfaber** [comment]
Hi, According to the paper the final loss (objective function) is calculated as a sum of multiple losses (Equations 3,4,5) Have a look at Figure 3 - There are multiple loss functions after each stage for both branches. To implement it in keras, we can pass a dictionary `loss` to a `compile` function (key: name of the output layer, value: name of the loss function for that layer) Similarly we pass a dictionary with loss weights where a key is name of the output layer and value is 1 because keras uses weighted sum to calculate final loss. That's why the model has quite a big number of outputs.

**Comment by legolas140** [comment]
Hi, Thank you very much for your reply. 
I could understand that the weighted sum needs to be calculated. My problem is that I have a new dataset(no coco), and I am trying to label it. I did't find this procedure in your code. After I check the question of liu6381810, I realize that maybe there is no code related to the label procedure in tensorflow, do I understand this correctly? 
And as to the final loss definition, I mean I didn't find the losses definition related to the Part Detection and PAF, like the Equation 6-15 in the paper. Is there any code related to equation 6-15? Thank you very much for your kind help!!

**Comment by michalfaber** [comment]
Hi @legolas140 
As far as I know, labeling used to be done manually using services like mechanical turk. Maybe there exist some tools which allow automating the process to some degree. Basically, you need to find x,y coordinates of body parts in the image - those are the annotations. Next step is preparing conf maps and paf which are feed into the network as labels. Implementation of eq 6-7 is here https://github.com/michalfaber/rmpe_dataset_transformer/blob/c138fb82b175d0a331b36e0fa8fc581913ac413f/DataTransformer.cpp#L331 and eq 8-9 herehttps://github.com/michalfaber/rmpe_dataset_transformer/blob/c138fb82b175d0a331b36e0fa8fc581913ac413f/DataTransformer.cpp#L349. Equations 10-15 are used for association of body parts in testing - I would say that this is the most complicated part of the pose estimation problem.

**Comment by legolas140** [comment]
Thank you very much for your reply!!!
In my own data set, I have already got the x,y coordinates of body parts, and I will use the code you mentioned to do the cons map and paf.
As to Equations 10-15, I mean I didn't find the related code reflecting these parts. Could you tell me where to find these codes (Equation 10-15)? so I could fully understand this code?
Thank you very much for your time!!

**Comment by michalfaber** [comment]
The related code starts from the line 68 https://github.com/michalfaber/keras_Realtime_Multi-Person_Pose_Estimation/blob/4d12c416a89cd2006446dd14af7dc88c4fa1dec1/demo_image.py#L68 Unfortunately my understanding of this part is still not very good.
